src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,572 @@
+                : GetFirstDiagnostic(child.Token);
+
+        private static EmbeddedDiagnostic? GetFirstDiagnostic(JsonToken token)
+            => GetFirstDiagnostic(token.LeadingTrivia) ?? token.Diagnostics.FirstOrNull() ?? GetFirstDiagnostic(token.TrailingTrivia);
+
+        private static EmbeddedDiagnostic? GetFirstDiagnostic(ImmutableArray<JsonTrivia> list)
+        {
+            foreach (var trivia in list)
+            {
+                var diagnostic = trivia.Diagnostics.FirstOrNull();
+                if (diagnostic != null)
+                    return diagnostic;
+            }
+
+            return null;
+        }
+
+        private ImmutableArray<JsonValueNode> ParseSequence()
+        {
+            try
+            {
+                _recursionDepth++;
+                StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
+                return ParseSequenceWorker();
+            }
+            finally
+            {
+                _recursionDepth--;
+            }
+        }
+
+        private ImmutableArray<JsonValueNode> ParseSequenceWorker()
+        {
+            using var _ = ArrayBuilder<JsonValueNode>.GetInstance(out var result);
+
+            while (ShouldConsumeSequenceElement())
+                result.Add(ParseValue());
+
+            return result.ToImmutable();
+        }
+
+        private JsonSeparatedList ParseCommaSeparatedSequence()
+        {
+            try
+            {
+                _recursionDepth++;
+                StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
+                return ParseCommaSeparatedSequenceWorker();
+            }
+            finally
+            {
+                _recursionDepth--;
+            }
+        }
+
+        private JsonSeparatedList ParseCommaSeparatedSequenceWorker()
+        {
+            using var _ = ArrayBuilder<JsonNodeOrToken>.GetInstance(out var result);
+            var allProperties = true;
+            while (ShouldConsumeSequenceElement())
+            {
+                var value = ParseValue();
+                allProperties = allProperties && value.Kind == JsonKind.Property;
+                result.Add(value);
+
+                // Try to consume a comma.  If we don't see one, consume an empty one as a placeholder. Create a
+                // diagnostic message depending on if we've seen only properties before this point.  If not, don't give
+                // a message about a missing comma.  Instead, we'll give a specific message that we didn't get a
+                // property when we expected one.
+                if (ShouldConsumeSequenceElement())
+                    result.Add(ConsumeToken(JsonKind.CommaToken, allProperties ? s_commaExpected : null));
+            }
+
+            return new JsonSeparatedList(result.ToImmutable());
+        }
+
+        private bool ShouldConsumeSequenceElement()
+        {
+            if (_currentToken.Kind == JsonKind.EndOfFile)
```

Comment:
nit: Can we make this a switch expression/statement?

Reply:
yup!